File: client/src/map3d/logistics-renderer.js
/**
 * 物流渲染器 — 运输车模型 + 路线可视化
 */
import * as THREE from 'three';
import { FACTION_COLORS_3D } from './constants.js';
import { toWorldXZ } from './world-space.js';
import { getHeightAt } from './terrain-generator.js';
import { stateStore } from '../bridge/state-store.js';

const logisticsMeshes = new Map(); // entityId → { cart, route, label }
let scene = null;

export function initLogisticsRenderer(sceneRef) { scene = sceneRef; }

/**
 * 刷新物流可视化
 */
export function refreshLogisticsVisuals() {
  const logistics = stateStore.logistics || {};

  // 移除已消失的路线
  for (const [entityId, data] of logisticsMeshes) {
    if (!logistics[entityId]) {
      removeLogisticsMesh(entityId);
    }
  }

  // 更新或创建
  for (const route of Object.values(logistics)) {
    let data = logisticsMeshes.get(route.entityId);

    if (!data) {
      data = createLogisticsMesh(route);
      if (data) {
        scene.add(data.group);
        logisticsMeshes.set(route.entityId, data);
      }
    }

    // 更新运输车位置
    if (data) {
      updateCartPosition(data, route);
    }
  }
}

/**
 * 创建物流路线可视化
 */
function createLogisticsMesh(route) {
  const group = new THREE.Group();
  group.userData = { type: 'logistics', entityId: route.entityId };

  // 路线线
  const nodes = stateStore.nodes;
  const pathPoints = [];
  if (route.pathNodeIds) {
    for (const nodeId of route.pathNodeIds) {
      const node = nodes[nodeId];
      if (node) {
        const { x, z } = toWorldXZ(node.x, node.y);
        pathPoints.push(new THREE.Vector3(x, getHeightAt(x, z) + 0.05, z));
      }
    }
  }

  let routeLine = null;
  if (pathPoints.length >= 2) {
    const lineGeo = new THREE.BufferGeometry().setFromPoints(pathPoints);
    const lineColor = route.mode === 'AUTO' ? 0x16A34A : 0x2563EB;
    const lineMat = new THREE.LineBasicMaterial({ color: lineColor, linewidth: 2, transparent: true, opacity: 0.6 });
    routeLine = new THREE.Line(lineGeo, lineMat);
    group.add(routeLine);
  }

  // 运输车
  const cart = createCartMesh(route);
  group.add(cart);

  return { group, cart, routeLine };
}

/**
 * 创建运输车模型
 */
function createCartMesh(route) {
  const group = new THREE.Group();

  // 车体
  const bodyGeo = new THREE.BoxGeometry(0.3, 0.15, 0.2);
  const bodyMat = new THREE.MeshStandardMaterial({ color: 0x8B7355, roughness: 0.8 });
  const body = new THREE.Mesh(bodyGeo, bodyMat);
  body.position.y = 0.12;
  group.add(body);

  // 车轮
  const wheelGeo = new THREE.CylinderGeometry(0.06, 0.06, 0.02, 8);
  const wheelMat = new THREE.MeshStandardMaterial({ color: 0x333333 });
  for (const [x, z] of [[-0.1, -0.12], [-0.1, 0.12], [0.1, -0.12], [0.1, 0.12]]) {
    const wheel = new THREE.Mesh(wheelGeo, wheelMat);
    wheel.rotation.x = Math.PI / 2;
    wheel.position.set(x, 0.06, z);
    group.add(wheel);
  }

  // 货物颜色标记
  const cargoGeo = new THREE.BoxGeometry(0.15, 0.08, 0.1);
  let cargoColor = 0x90EE90; // FOOD=绿
  if (route.cargoType === 'IRON') cargoColor = 0x808080;
  if (route.cargoType === 'AMMO') cargoColor = 0xDC2626;
  const cargoMat = new THREE.MeshStandardMaterial({ color: cargoColor });
  const cargo = new THREE.Mesh(cargoGeo, cargoMat);
  cargo.position.y = 0.23;
  group.add(cargo);

  group.castShadow = true;
  return group;
}

/**
 * 更新运输车位置
 */
function updateCartPosition(data, route) {
  const nodes = stateStore.nodes;
  const cart = data.cart;

  // 简化：使用 route 的当前边和进度
  if (route.currentEdgeId && stateStore.edges[route.edgeId || route.currentEdgeId]) {
    const edge = stateStore.edges[route.currentEdgeId];
    if (edge) {
      const srcNode = nodes[edge.sourceNodeId];
      const tgtNode = nodes[edge.targetNodeId];
      if (srcNode && tgtNode) {
        const src = toWorldXZ(srcNode.x, srcNode.y);
        const tgt = toWorldXZ(tgtNode.x, tgtNode.y);
        const t = route.edgeProgress || 0;
        const wx = src.x + (tgt.x - src.x) * t;
        const wz = src.z + (tgt.z - src.z) * t;
        const wy = getHeightAt(wx, wz);
        cart.position.set(wx, wy + 0.05, wz);
      }
    }
  }
}

function removeLogisticsMesh(entityId) {
  const data = logisticsMeshes.get(entityId);
  if (data) {
    scene.remove(data.group);
    data.group.traverse((obj) => {
      if (obj.geometry) obj.geometry.dispose();
      if (obj.material) obj.material.dispose();
    });
    logisticsMeshes.delete(entityId);
  }
}

/** 清理 */
export function clearAllLogistics() {
  for (const [id, data] of logisticsMeshes) {
    scene.remove(data.group);
    data.group.traverse((obj) => {
      if (obj.geometry) obj.geometry.dispose();
      if (obj.material) obj.material.dispose();
    });
  }
  logisticsMeshes.clear();
}

File: U3D/open-world-game-unity/Assets/Scripts/Rendering/LogisticsRenderer.cs
using System.Collections.Generic;
using GameState;
using UnityEngine;
using Rendering;

/// <summary>
/// 物流渲染器 — 运输车模型 + 路线可视化
/// 移植自 client/src/map3d/logistics-renderer.js
/// </summary>
namespace Rendering
{
    public class LogisticsRenderer : MonoBehaviour
    {
        // entityId → { root, cart }
        private readonly Dictionary<int, LogisticsVisual> _logisticsObjects = new();
        private TerrainGenerator _terrain;

        private class LogisticsVisual
        {
            public GameObject root;
            public GameObject cart;
        }

        private void Awake()
        {
            _terrain = FindObjectOfType<TerrainGenerator>();
        }

        /// <summary>刷新所有物流可视化</summary>
        public void RefreshLogistics(Dictionary<int, LogisticsComponent> logistics)
        {
            if (logistics == null) return;

            // 移除已消失的
            var toRemove = new List<int>();
            foreach (var (id, _) in _logisticsObjects)
            {
                if (!logistics.ContainsKey(id))
                    toRemove.Add(id);
            }
            foreach (var id in toRemove)
                RemoveLogistics(id);

            // 更新或创建
            foreach (var (id, route) in logistics)
            {
                if (!_logisticsObjects.ContainsKey(id))
                {
                    var vis = CreateLogisticsVisual(route);
                    if (vis != null)
                    {
                        vis.root.transform.SetParent(transform, false);
                        _logisticsObjects[id] = vis;
                    }
                }

                if (_logisticsObjects.TryGetValue(id, out var existing))
                {
                    UpdateCartPosition(existing.cart, route);
                }
            }
        }

        /// <summary>创建物流路线可视化</summary>
        private LogisticsVisual CreateLogisticsVisual(LogisticsComponent route)
        {
            var root = new GameObject($"Logistics_{route.EntityId}");

            var nodes = global::GameApp.Instance?.State?.Nodes;
            var pathPoints = new List<Vector3>();

            if (route.PathNodeIds != null && nodes != null)
            {
                foreach (var nodeId in route.PathNodeIds)
                {
                    if (nodes.TryGetValue(nodeId, out var node))
                    {
                        var xz = CoordinateUtils.ToWorldXZ(node.X, node.Y);
                        float y = (_terrain?.GetHeightAt(xz.x, xz.y) ?? 0f) + 0.05f;
                        pathPoints.Add(new Vector3(xz.x, y, xz.y));
                    }
                }
            }

            // 路线线
            if (pathPoints.Count >= 2)
            {
                var lineObj = new GameObject("RouteLine");
                lineObj.transform.SetParent(root.transform, false);

                var lr = lineObj.AddComponent<LineRenderer>();
                lr.positionCount = pathPoints.Count;
                lr.SetPositions(pathPoints.ToArray());
                lr.startWidth = 0.05f;
                lr.endWidth = 0.05f;
                Color lineColor = route.Mode == "AUTO"
                    ? new Color(0.09f, 0.63f, 0.29f, 0.6f)
                    : new Color(0.15f, 0.39f, 0.92f, 0.6f);
                lr.startColor = lineColor;
                lr.endColor = lineColor;
                lr.material = MaterialCache.GetUnlit(lineColor);
                lr.material.color = lineColor;
                lr.shadowCastingMode = UnityEngine.Rendering.ShadowCastingMode.Off;
            }

            // 运输车
            var cart = CreateCart(route);
            if (cart != null)
                cart.transform.SetParent(root.transform, false);

            return new LogisticsVisual { root = root, cart = cart };
        }

        /// <summary>创建运输车模型</summary>
        private static GameObject CreateCart(LogisticsComponent route)
        {
            var cart = new GameObject("Cart");

            // 车体
            var bodyObj = GameObject.CreatePrimitive(PrimitiveType.Cube);
            bodyObj.name = "Body";
            DestroyImmediate(bodyObj.GetComponent<Collider>());
            bodyObj.transform.SetParent(cart.transform, false);
            bodyObj.transform.localScale = new Vector3(0.3f, 0.15f, 0.2f);
            bodyObj.transform.localPosition = new Vector3(0, 0.12f, 0);
            bodyObj.GetComponent<MeshRenderer>().sharedMaterial = MaterialCache.GetLit(new Color(0.54f, 0.45f, 0.33f));

            // 车轮
            var wheelObj = GameObject.CreatePrimitive(PrimitiveType.Cylinder);
            wheelObj.name = "Wheel";
            DestroyImmediate(wheelObj.GetComponent<Collider>());
            wheelObj.transform.SetParent(cart.transform, false);
            wheelObj.transform.localScale = new Vector3(0.06f, 0.02f, 0.06f);
            wheelObj.transform.localRotation = Quaternion.Euler(90, 0, 0);
            var wheelMat = MaterialCache.GetLit(new Color(0.2f, 0.2f, 0.2f));

            foreach (var (wx, wz) in new[] { (-0.1f, -0.12f), (-0.1f, 0.12f), (0.1f, -0.12f), (0.1f, 0.12f) })
            {
                var w = Object.Instantiate(wheelObj, cart.transform);
                w.transform.localPosition = new Vector3(wx, 0.06f, wz);
                w.GetComponent<MeshRenderer>().sharedMaterial = wheelMat;
            }
            Object.DestroyImmediate(wheelObj);

            // 货物
            var cargoObj = GameObject.CreatePrimitive(PrimitiveType.Cube);
            cargoObj.name = "Cargo";
            DestroyImmediate(cargoObj.GetComponent<Collider>());
            cargoObj.transform.SetParent(cart.transform, false);
            cargoObj.transform.localScale = new Vector3(0.15f, 0.08f, 0.1f);
            cargoObj.transform.localPosition = new Vector3(0, 0.23f, 0);
            Color cargoColor = route.CargoType switch
            {
                "IRON" => new Color(0.5f, 0.5f, 0.5f),
                "AMMO" => new Color(0.86f, 0.15f, 0.15f),
                _ => new Color(0.56f, 0.93f, 0.56f), // FOOD = 绿
            };
            cargoObj.GetComponent<MeshRenderer>().sharedMaterial = MaterialCache.GetLit(cargoColor);

            return cart;
        }

        /// <summary>更新运输车位置</summary>
        private void UpdateCartPosition(GameObject cart, LogisticsComponent route)
        {
            if (cart == null || _terrain == null) return;

            var store = global::GameApp.Instance?.State;
            var nodes = store?.Nodes;
            var edges = store?.Edges;

            if (edges == null || nodes == null) return;

            string edgeId = route.CurrentEdgeId ?? route.CurrentEdgeId;
            if (!string.IsNullOrEmpty(edgeId) && edges.TryGetValue(edgeId, out var edge))
            {
                if (nodes.TryGetValue(edge.SourceNodeId, out var src) &&
                    nodes.TryGetValue(edge.TargetNodeId, out var tgt))
                {
                    var srcW = CoordinateUtils.ToWorldXZ(src.X, src.Y);
                    var tgtW = CoordinateUtils.ToWorldXZ(tgt.X, tgt.Y);
                    float progress = Mathf.Clamp01(route.EdgeProgress);

                    float wx = srcW.x + (tgtW.x - srcW.x) * progress;
                    float wz = srcW.y + (tgtW.y - srcW.y) * progress;
                    float wy = _terrain.GetHeightAt(wx, wz) + 0.05f;

                    cart.transform.position = new Vector3(wx, wy, wz);
                }
            }
        }

        private void RemoveLogistics(int entityId)
        {
            if (_logisticsObjects.TryGetValue(entityId, out var vis))
            {
                Destroy(vis.root);
                _logisticsObjects.Remove(entityId);
            }
        }

        /// <summary>清理所有物流对象</summary>
        public void ClearAllLogistics()
        {
            foreach (var (id, vis) in _logisticsObjects)
                Destroy(vis.root);
            _logisticsObjects.Clear();
        }

        private void OnDestroy() => ClearAllLogistics();
    }
}
